lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,3148 @@
+
+      while (parent->getParentOp() != defOp->getParentOp())
+        parent = parent->getParentOp();
+
+      freeClassicals[parent].insert(shadow_node);
+
+      return DependencyNode::DependencyEdge{shadow_node, resultidx};
+    }
+
+    return DependencyNode::DependencyEdge{dnode, resultidx};
+  }
+
+  /// Cleans up semi-constructed dependency graph when backing out of running
+  /// DependencyAnalysis because of an encountered error
+  void clean() {
+    // TODO: clean up nodes
+    // Adam: can use perOps, have to be careful about nested nodes though
+  }
+
+  unsigned getNumVirtualAllocs() { return vallocs; }
+};
+
+struct DependencyAnalysisPass
+    : public cudaq::opt::impl::DependencyAnalysisBase<DependencyAnalysisPass> {
+  using DependencyAnalysisBase::DependencyAnalysisBase;
+
+  /// DependencyAnalysis constructs a data structure representing the
+  /// quake AST, performs several analyses/optimizations, and then generates
+  /// a new quake AST based on the resulting data structure.
+  ///
+  /// First, the quake AST is walked, constructing a DependencyBlock for the
+  /// body of every kernel function.
+  ///
+  /// Next, virtual qubit allocations are lowered to the inner-most scope in
+  /// which they are used (see `contractAllocsPass`). This works outside-in.
+  /// Lowering virtual qubit allocations opens up more qubit reuse opportunities
+  /// within inner scopes, and gives the remaining optimizations more
+  /// flexibility.
+  ///
+  /// Then, an inside-out analysis/optimization pass is performed, assigning
+  /// physical qubits to virtual wires, and lifting common optimizations. This
+  /// inside-out analysis/optimization pass works as follows:
+  /// - Step 1: `if`s inside blocks are resolved (hence inside-out), once an
+  /// `if` is resolved, the information contained by it will not be changed
+  /// (with the possible exception of re-indexing physical qubits).
+  /// - Step 2: blocks are resolved.
+  /// * First the nodes inside them are scheduled, assigning a cycle to every
+  ///   node such that each node is scheduled after all of its dependencies, and
+  ///   before all of its successors.
+  /// * Then, physical qubits are allocated and assigned to virtual wires based
+  ///   on lifetime information (i.e., which cycles the virtual wire is used
+  ///   in). This algorithm treats `if`s as "solid rectangles", where all qubits
+  ///   in use anywhere in the `if` are considered in use for the entire `if` by
+  ///   the parent scope. In other words, the lifetime analysis does not look
+  ///   inside `if`s.
+  /// - Step 3: Once its blocks are resolved, then the parent `if` is resolved.
+  /// * First, the allocations from the two inner blocks are combined/matched
+  ///   (respecting reuse within the blocks but with re-indexing allowed).
+  /// * Second, equivalent operations at the beginning/end of the then and else
+  ///   blocks are lifted to the parent context, before/after the `if`.
```

Comment:
I'm curious how the virtual qids interact with the determination of equivalence. If the two arms use different qid numberings, that doesn't necessarily mean that the prefixes aren't common, for example.

Reply:
Virtual wires only allocated inside the `if` will be assigned physical qubits by the analysis, so equivalence should only look at the physical qubit assigned to those. For virtual wires allocated in some parent scope, we look at the qid to ensure that it is the same virtual wire.

Reply:
> we look at the qid
Which is the qid assigned in the function scope?

Reply:
Well, the qid assigned in the the parent scope (which may be the function scope).